src/Build/BackEnd/Node/OutOfProcServerNode.cs

```diff
@@ -0,0 +1,516 @@
+
+            // configure console output redirection
+            var oldOut = Console.Out;
+            var oldErr = Console.Error;
+
+            using var outWriter = new RedirectConsoleWriter(text =>
+            {
+                SendPacket(new ServerNodeConsoleWrite(text, 1));
+            });
+            using var errWriter = new RedirectConsoleWriter(text =>
+            {
+                SendPacket(new ServerNodeConsoleWrite(text, 2));
+            });
+
+            Console.SetOut(outWriter);
+            Console.SetError(errWriter);
+
+            var (exitCode, exitType) = _buildFunction(command.CommandLine);
+
+            Console.SetOut(oldOut);
+            Console.SetError(oldErr);
+            outWriter.Dispose();
+            errWriter.Dispose();
+
+            // On Windows, a process holds a handle to the current directory,
+            // so reset it away from a user-requested folder that may get deleted.
+            NativeMethodsShared.SetCurrentDirectory(BuildEnvironmentHelper.Instance.CurrentMSBuildToolsDirectory);
+
+            var response = new ServerNodeResponse(exitCode, exitType);
+            SendPacket(response);
+
+            _shutdownReason = NodeEngineShutdownReason.BuildCompleteReuse;
+            _shutdownEvent.Set();
+        }
+
+        // TODO: unit tests
+        internal class RedirectConsoleWriter : StringWriter
+        {
+            private readonly string _newLineString;
+
+            private readonly Action<string> _writeCallback;
+
+            public RedirectConsoleWriter(Action<string> writeCallback)
+            {
+                _newLineString = new String(CoreNewLine);
+                _writeCallback = writeCallback;
+            }
+
+            private void MaybeFlushCaptured(bool force)
+            {
+                if (force || GetStringBuilder().Length > 200)
+                {
+                    Flush();
+                }
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                if (disposing)
+                {
+                    Flush();
+                }
+
+                base.Dispose(disposing);
+            }
+
+            public override void Flush()
+            {
+                var sb = GetStringBuilder();
+                var captured = sb.ToString();
+                sb.Clear();
+                _writeCallback(captured);
+
+                base.Flush();
+            }
+
+            public override void Write(char value)
+            {
+                base.Write(value);
+                MaybeFlushCaptured(false);
+            }
+
+            public override void Write(char[] buffer, int index, int count)
+            {
+                base.Write(buffer, index, count);
+                MaybeFlushCaptured(buffer.SequenceEqual(CoreNewLine));
+            }
+
+            public override void Write(string? value)
+            {
+                if (value is null)
```

Comment:
super-nit: Use `IsNullOrEmpty` to skip empty strings as well?

Reply:
Custom `Write` overrides were removed.